### ISSUE TITLE
Consider allowing implicit yield in lists and arrays

### ISSUE THREAD

**Issue by dsyme** [fsharp/fslang-suggestions#643]
F# list and arrays (but not sequences and computation expressions) allow both data and computations that produce data, e.g.

```fsharp
    [ 1; 2 ]
    [ yield 1; yield 2 ]
    [ yield 1; yield 2; match a with A -> yield 1 | B -> () ]
```

but not

```fsharp
    [ 1; 2; match a with A -> yield 1 | B -> () ]
```

This means that adding a single computation to an otherwise simple data list requires you to add ``yield`` everywhere.  This particularly bites computed views in Fable/Elmish.

It is possible we could be more tolerant here and interpret expressions in the list which are not unit-typed and are at the fringes of control-flow expressions as an implicit yield. This would allow

```fsharp
    [ 1; 2; match a with A -> 1 | B -> () ]
```
Note that the ``yield`` would, I think, still be required in the case of in element that is a control construct ``match``, ``let``, ``if`` etc.

It's possible that making such a more liberal interpretation can confuse people in other ways or be a breaking change.  For example, 
1. the branches of the ``match`` no longer match in type - one is unit-typed and the other is not.  This means the expression can't be extracted into a function  However this is already  the case - if the expression syntax contains ``yield`` then it is part of a generator, and is not an expression. 
2. existing code might already contain non-unit-valued items that are currently being ignored (with a warning).  In this new interpretation these would now be yielded.
3. Lists of unit values would not satisfy this

We would also need to consider whether the construct also applies to sequence expressions (implicit yield) and computation expressions (implicit yield/return).

### Example

Consider this:
```
                    div [ClassName ("form-group has-feedback " + authorStatus) ] [
                         yield div [ClassName "input-group"][
                             yield span [ClassName "input-group-addon"] [span [ClassName "glyphicon glyphicon-user"] [] ]
                             yield input [
                                     Key ("Author_" + model.NewBookId.ToString())
                                     HTMLAttr.Type "text"
                                     Name "Author"
                                     DefaultValue model.NewBook.Authors
                                     ClassName "form-control"
                                     Placeholder "Please insert authors"
                                     Required true
                                     OnChange (fun (ev:React.FormEvent) -> dispatch (AuthorsChanged !!ev.target?value))]
                             match model.AuthorsErrorText with
                             | Some e -> yield span [ClassName "glyphicon glyphicon-remove form-control-feedback"] []
                             | _ -> ()
                         ]
                         match model.AuthorsErrorText with
                         | Some e -> yield p [ClassName "text-danger"][str e]
                         | _ -> ()
                    ]
```
Note that the presence of one ``match`` requires adding lots of ``yield``.  This would become:
```
                    div [ClassName ("form-group has-feedback " + authorStatus) ] [
                         div [ClassName "input-group"][
                             span [ClassName "input-group-addon"] [span [ClassName "glyphicon glyphicon-user"] [] ]
                             input [
                                 Key ("Author_" + model.NewBookId.ToString())
                                 HTMLAttr.Type "text"
                                 Name "Author"
                                 DefaultValue model.NewBook.Authors
                                 ClassName "form-control"
                                 Placeholder "Please insert authors"
                                 Required true
                                 OnChange (fun (ev:React.FormEvent) -> dispatch (AuthorsChanged !!ev.target?value))]
                             match model.AuthorsErrorText with
                             | Some e -> yield span [ClassName "glyphicon glyphicon-remove form-control-feedback"] []
                             | _ -> ()
                         ]
                         match model.AuthorsErrorText with
                         | Some e -> yield p [ClassName "text-danger"][str e]
                         | _ -> ()
                    ]
```

## Pros and Cons 

The advantages of making this adjustment to F# are clearer code in the common case

The disadvantages of making this adjustment to F# are 
* it's a corner-case breaking change 
* it would require a type-directed rule
* ``yield!`` or ``return!`` would still be required to yield another computed thing.  Forgetting to add ``yield!`` or ``return!`` might be even more confusing than it already is

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

cc @forki

## Affidavit (please submit!)

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design

**Comment by dsyme** [comment]
Here's another example.  
```fsharp

let view onLogout (model:Model) =
    div [ centerStyle "row" ] [
          yield viewLink Page.Home "Home"
          if model <> None then
              yield viewLink Page.WishList "Wishlist"
          if model = None then
              yield viewLink Page.Login "Login"
          else
              yield buttonLink "logout" onLogout [ str "Logout" ]
        ]
```
which could become:
```fsharp

let view onLogout (model:Model) =
    div [ centerStyle "row" ] [
          viewLink Page.Home "Home"
          if model <> None then
              viewLink Page.WishList "Wishlist"
          if model = None then
              viewLink Page.Login "Login"
          else
              buttonLink "logout" onLogout [ str "Logout" ]
        ]
```
That would again require a type-directed rule (here we are eliding ``yield`` even from the if-then-else control construct).

**Comment by Rickasaurus** [comment]
Could this be made to work in all computation expression contexts? Would we be able to enable or disable it, maybe based on an annotation or something?

**Comment by dsyme** [comment]
@rickasaurus Yes it could be made to work.  Yes it could be controlled by an annotation 

We would have to decide annotations for ``seq`` and ``async`` - my intuition is that it would need to be on for ``seq`` for parity with ``list`` and ``async

I'm much less sure about ``async`` - an explicit ``return`` is currently needed for async but this is actually a point of friction when changing code from sync to async.  However the use of unit-typed side-effecting code is very, very common in Async code, and of course in Async you aren't defining a list of items anyway, and only the final item can even have "return".

**Comment by Rickasaurus** [comment]
Spot on, Async was exactly the case I was thinking about.

**Comment by dsyme** [comment]
@Rickasaurus It's possible that only applying the rule to ``yield`` would make sense

**Comment by tpetricek** [comment]
This looks like a pretty small change that would make a lot of my Fable code nicer 👍

**Comment by rmunn** [comment]
I like this:
```fsharp
[ 1; 2; match a with A -> yield 1 | B -> () ]
```
and I would also like this:
```fsharp
[ 1; 2; match a with A -> 1 | B -> 2 ]
```
but I don't like this:
```fsharp
[ 1; 2; match a with A -> 1 | B -> () ]
```

The type mismatch in that last example bothers me, since it appears to break the usual rule of "every branch of an `if` expression or a `match` expression must have the same type". I think that not requiring `yield` in the first two examples makes a lot of sense and will be easy for beginners to learn, but the third example feels like breaking the rules without a good enough reason.

**Comment by dsyme** [comment]
@rmunn what about the if/then and if/then/else cases?

**Comment by forki** [comment]
Would be very happy to have this.

**Comment by rmunn** [comment]
The if/then/else case is pretty obvious: if branches are the same type, then `yield` can be omitted, but if the branches are two different types, `yield` is required. OTOH, when it's just an if/then (where the `else` clause is implicit), it doesn't bother me to let the `then` clause have an implicit yield. E.g., this code feels fine to me, because the `else` clause is out of sight, out of mind:

```fsharp
  [ "1"
    "2"
    if x % 3 = 0 then
        "buzz"
  ]
```

But the following bothers me a little:

```fsharp
  [ "1"
    "2"
    if x % 3 = 0 then
        "buzz"  // I'd prefer `yield "buzz"` here
    else
        printfn "No result here"
  ]
```

OTOH, the fact that adding an `else` clause with `printfn`, or a unit-returning logging function, makes you have to suddenly add a `yield` is also a bit bothersome, so maybe I can live with the `unit` exception for implied yields. And if I can live with it for `if` expressions, I can probably live with it for `match` expressions as well. I still don't prefer it, but I can live with it, especially if the `unit` value is being returned by the `ignore` function.

In other words, although this example bothered me:

```fsharp
[ 1; 2; match a with A -> 1 | B -> () ]
```

This example bothers me less, somehow:

```fsharp
[ 1; 2; match a with A -> 1 | B -> 2 |> ignore ]
```

This is not a consistent position; it's just my intuition, and I'm trying to work out why my intuition is inconsistent here. Both of those two examples have a `unit` value in the second branch, so why am I bothered by the first and not the second? I don't know, and the more I look at this, the less I'm really bothered by the first example. I think I can live with it; I still don't like it all that much, but I'm probably at a -0 on the idea now rather than a -1. (That is, the idea of allowing implicit yield on a branch where a different branch returns `unit`).

**Comment by dsyme** [comment]
@rmunn that's the exact path of reasoning I travelled down too :)

**Comment by jcmrva** [comment]
Is something like this possible?
```
[ 1; 2; match a with A -> 1 | B -> _ |> ignore ]
```
Or even...
```
[ 1; 2; match a with A -> 1 | B -> _ ]
```
Defining an arbitrary value just to ignore it seems weird.

**Comment by rmunn** [comment]
A real usage scenario would far more likely be:
```fsharp
[ 1; 2; match a with A -> 1 | B -> log "An error occurred" ]
```
where `log` is a function of type `string -> unit`. So I don't think there's a need for `_` in this scenario, because most of the time the B branch of the match expression isn't going to look like `B -> ()`.

**Comment by chadunit** [comment]
This would be great! 👍 

Speaking of computation noise in Elmish views, I would like to ask if there's any merit -- either as part of this or as a separate suggestion -- to retaining implicit upcasts when adding a computation.

For example this with implicit upcasts:
```
a [
  Href "#" // HTMLAttr implicit upcast to IHTMLProp 
  OnClick handlerFn // DOMAttr implicit upcast to IHTMLProp
] []

```

Currently becomes this with explicit upcasts:
```
a [
  yield upcast Href "#" // Now requires explicit upcast HTMLAttr :> IHTMLProp
  if someCondition then
    yield upcast OnClick handlerFn // Now requires explicit upcast DOMAttr :> IHTMLProp
] []
```

It appears this proposal would eliminate the yields, but I assume still require the upcasts:
```
a [
  upcast Href "#" // explicit upcast
  if someCondition then
    upcast OnClick handlerFn // explicit upcast
] []
```

**Comment by psiLearn** [comment]
Right now I would write:

``

**Comment by dsyme** [comment]
@chadunit You are correct that the upcast problem is separate, but should likely be considered as part of this

**Comment by jwosty** [comment]
FWIW, @rmunn's example of

    [ "1"
      "2"
      if x % 3 = 0 then
          "buzz" ]

kinda bothers me. My intuition says the `yield` should be explicit here. What if you call a function for side effects in that branch, but you don't realize it returns `int` and not `unit`? You wouldn't get a compiler error or even a warning, and the expression would implicitly `yield` the return value into the list.

Maybe that would just be a crazy corner case, but I wouldn't have fun debugging that if I ran into it. One might be more likely to fall into that trap when working with C interop code.

**Comment by xperiandri** [comment]
What about just prepending list with the yield
``` F#
yield [ 1; 2; match a with A -> 1 | B -> () ]
```

**Comment by isaacabraham** [comment]
This certainly simplifies a lot of code.

However I do worry about this changing - what seems to me - to be a kind of fundamental behaviour of if / then expressions in F# i.e. each branch returns some value that typically gets assigned to some binding etc. - here they just appear to go into "nothing" and magically get yielded out. On the other hand, I've seen beginners to F# and Elmish get bitten by this lots of times.

Maybe this would be doing "the right thing" and it's a case of unlearning how I think about if / then expressions in F#. I guess this just means that within the context of lists / arrays / sequences that yield is no longer required, that's it?

**Comment by chadunit** [comment]
Just FYI to anyone interested, the incidental upcast issue I asked about [earlier in this thread](https://github.com/fsharp/fslang-suggestions/issues/643#issuecomment-362926508) was already implemented and shipped separately, see [RFC FS-1055](https://github.com/fsharp/fslang-design/blob/master/FSharp-4.5/FS-1055-subsumption-for-yield-in-sequence-expression.md). 🚢 😄

**Comment by cartermp** [comment]
Closing this since it's done

**Comment by gusty** [comment]
I was playing a bit with this new feature.

There's something that I do not understand, why do I need to specify a `Zero` method, even when it's not called at all?

Here's an example:

      myCE {1; 2; 3} ;;
      ------------^

    stdin(4,12): error FS0708: This control construct may only be used if the computation expression 
    builder defines a 'Zero' method }

I can understand that if do:

      myCE {1; 2; 3; ()} ;;

it complains, but why require it in the first case, where it's not been used at all?
